### ISSUE TITLE
Kubelet stop posting node status(node notReady) when change the date time to an earlier time

### ISSUE THREAD

**Issue by xialonglee** [kubernetes/kubernetes#54130]
**Is this a BUG REPORT or FEATURE REQUEST?**:

 kind bug

**What happened**:

When change the date time on node to an earlier time: 
e.g. change the date 2017-10-18 14:00  -> 2017-10-18 11:00
will cause kubelet **stop posting its node's status for 3 hours**.

This will cause the node status become **notReady** from master side.

How long we should wait for kubelet posting node's status again depends on how long the distance of time we changed to the original time (e.g. if we change to time to half hour earlier, kubelet will stop posting status for half an hour).

I check the code and debug, finally found that when posting node status, kubelet will invoke `kl.kubeClient.Core().Nodes().Get(string(kl.nodeName), opts)` to retrieve node info from apiserver.
When modify the time to earlier, it stuck here, and I check the codes further in `vendor/k8s.io/client-go/rest/request.go` and found that it stuck in invokation of `r.tryThrottle()`.

There's some mechanism will exec `time.sleep` so that the kubelet will stop posting node's status. See it below:

vendor/github.com/juju/ratelimit/ratelimit.go
```
/ Wait takes count tokens from the bucket, waiting until they are
// available.
func (tb *Bucket) Wait(count int64) {
	if d := tb.Take(count); d > 0 {
		time.Sleep(d)
	}
}
```

Restart the kubelet service will fix.

**What you expected to happen**:
for some reason, we may synchronize the time, so there has the possibility that change the time to an earlier time, but we don't want restart kubelet service and hope it will continue posting the node's status.

**How to reproduce it (as minimally and precisely as possible)**:
When setup the cluster normally， just change the node‘s date time to an earlier will cause the node become `notReady`.

PS: change the time to later time will not cause this problem.

**Environment**:
- Kubernetes version (use `kubectl version`):  GitVersion:"v1.6.9+a3d1dfa6f4335"
- Others:

**Comment by xialonglee** [comment]
@kubernetes/sig-node

**Comment by xialonglee** [comment]
@kubernetes/sig-node-bugs
/sig node

**Comment by bruceauyeung** [comment]
If i understand correctly, `juju/ratelimit` records requests' startTime and uses `now.sub(startTime)`(or something like this) to determine if request should be sent. if system time is changed ,time-measuring operations like `time.Sub(another)` will not work as expected.
~~~
start := time.Now()
... operation that takes 20 milliseconds ...
t := time.Now()
elapsed := t.Sub(start)
~~~
but Golang 1.9 brings [Transparent Monotonic Time support](https://golang.org/doc/go1.9#monotonic-time), so i think upgrading Golang to 1.9 may fix this bug.

**Comment by mbohlool** [comment]
cc @yliaog

**Comment by lavalamp** [comment]
Dup of #54236?

**Comment by yliaog** [comment]
Yes, Go 1.9 [Transparent Monotonic Time support](https://golang.org/doc/go1.9#monotonic-time) should fix it.

In the meantime, a quick fix is to add a timeout to the Wait. i.e. instead of using [Wait](https://github.com/kubernetes/kubernetes/blob/6ddaee5ed12f0a6ad3eca2ea3ef679180cc1d348/vendor/github.com/juju/ratelimit/ratelimit.go#L141), use [WaitMaxDuration](https://github.com/kubernetes/kubernetes/blob/6ddaee5ed12f0a6ad3eca2ea3ef679180cc1d348/vendor/github.com/juju/ratelimit/ratelimit.go#L152)

In the following code:
[t.limiter.Wait(1)](https://github.com/kubernetes/kubernetes/blob/6ddaee5ed12f0a6ad3eca2ea3ef679180cc1d348/staging/src/k8s.io/client-go/util/flowcontrol/throttle.go#L87)

**Comment by xialonglee** [comment]
@lavalamp 
no,these problem are under different code and maybe are different mechanism.

**Comment by xialonglee** [comment]
@bruceauyeung @yliaog 
I would try build `kubelet` with Go 1.9 later, thx!

**Comment by xialonglee** [comment]
I tested it that using Go 1.9 to build `kubelet` and Go 1.9 fix this issue.

**Comment by dchen1107** [comment]
I am glad using Go 1.9 can fix the issue. But I don't understand why we have concern to restart Kubelet service to workaround the issue at the first place?

**Comment by bruceauyeung** [comment]
I think restarting kubelet service is not a good option

**Comment by yliaog** [comment]
i think the question boils down to how much disruption restarting kubelet would cause. In some k8s deployment, kubelet restart would cause the containers on the node to be restarted. not sure if this is the case for @bruceauyeung though.

**Comment by liggitt** [comment]
Closing this issue as fixed since 1.9 is used to build now
/close
